syntax = "proto3";

package indigo.igpackgen.pb;

/*************************************************\
 *         Indigo Asset Pack Plan Proto          *
 *                                               *
 * Declarative format for loading and processing *
 *  raw asset source files, and compiling them   *
 *  into asset packs that can later be used by   *
 *  the Indigo engine.                           *
\*************************************************/

message Vec3f {
  float x = 1;
  float y = 2;
  float z = 3;
}

// Rotation, expressed as an axis and angle (to build a quaternion)
message Rotation {
  float x = 1;
  float y = 2;
  float z = 3;
  
  float angle = 4;
}

message CopyWgslSourceAction {
  enum ShaderType {
    UNKNOWN = 0;
    VERTEX = 1;
    FRAGMENT = 2;
    COMPUTE = 3;
  }

  string input_file_path = 1;
  ShaderType shader_type = 2;
  string entry_point = 3;

  string igasset_name = 4;
  bool strip_comments_and_such = 5;
}

message DracoConversionParams {
  int32 compression_speed = 1;
  int32 decompression_speed = 2;
}

message AssimpToStaticDracoGeoAction {
  string input_file_path = 1;
  repeated string assimp_mesh_names = 2;
  DracoConversionParams draco_params = 3;

  string mesh_igasset_name = 4;
}

message AssembleRecastNavMeshAction {
  // XZ cell size - param "cs" in Recast API calls
  float cell_size = 1;

  message BB {
    float min_x = 1;
    float min_y = 2;
    float min_z = 3;
    float max_x = 4;
    float max_y = 5;
    float max_z = 6;
  }

  // Bounding box override - params bim/bmax in Recast API calls
  BB bb_override = 2;

  message AssimpGeoDef {
    string assimp_file_name = 1;
    string assimp_mesh_name = 2;

    Vec3f position = 3;
    Rotation rotation = 4;
    Vec3f scale = 5;
  }

  message RecastOp {
    oneof op {
      AssimpGeoDef include_assimp_geo = 1;
      AssimpGeoDef exclude_assimp_geo = 2;
    }
  }

  // An ordered list of Recast construction operations to apply in order
  //  to build the Recast navmesh.
  repeated RecastOp recast_build_ops = 3;
}

message SingleAction {
  oneof request {
    CopyWgslSourceAction copy_wgsl_source = 1;
    AssimpToStaticDracoGeoAction assimp_to_static_draco_geo = 2;
  }
}

message SingleIgpackPlan {
  string asset_pack_file_path = 1;
  repeated SingleAction actions = 2;
}

message IgpackGenPlan {
  repeated SingleIgpackPlan plan = 1;
}